onnxruntime/core/graph/graph_utils.cc

```diff
@@ -161,34 +161,30 @@ static void RemoveGraphEdges(Graph& graph, const std::vector<GraphEdge>& edges)
 /** Given a graph, a list of edges, and a NodeArg name, checks if each of the edges provides an implicit input
     to a subgraph. If so, it checks if there is no clash of the given NodeArg name in each of the subgraphs. 
     This is important when removing a node with this NodeArg as input. */
-bool CanUpdateImplicitInputNameInSubgraphs(Graph& graph,
+bool CanUpdateImplicitInputNameInSubgraphs(const Graph& graph,
                                            const std::vector<GraphEdge>& output_edges,
-                                           const std::string& input_arg_name) {
+                                           const std::string& new_arg_name) {
   for (const auto& output_edge : output_edges) {
     if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
-      Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
-      if (!CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, input_arg_name)) {
-        LOGS_DEFAULT(WARNING) << " Implicit input name " << input_arg_name
-                              << " cannot be safely updated in one of the subgraphs.";
+      const Node& output_edge_node = *graph.GetNode(output_edge.dst_node);
+      if (!CanUpdateImplicitInputNameInSubgraph(output_edge_node, output_edge.arg_name, new_arg_name)) {
+        LOGS_DEFAULT(WARNING) << " Implicit input name " << output_edge.arg_name
+                              << " cannot be safely updated to " << new_arg_name << " in one of the subgraphs.";
         return false;
       }
     }
   }
+
   return true;
 }
 
-/** Removes a node with a single incoming node. */
+/** Removes a node with a single incoming node and replaces its output with the incoming node's output. */
```

Comment:
Maybe depends on how you view the edges. I was coming from the perspective that the node was producing output named 'outputB', and that is being replaced with the output named 'outputA' from the incoming node. We update the downstream node/s to update their input defs to replace outputB with outputA and that is what the edge is based on. 

However saying that just made me notice it only creates a new edge and doesn't update the input defs of the downstream node/s which I believe is incorrect based on Graph::RemoveNode expecting that the NodeArg values at each end of the edge refer to the same instance. 

I'll update a few things to hopefully clarify and make more correct.

Reply:
There is indeed a moment when we remove a node and then add the edge between its input and output nodes that the nodeargs are not the same, but there is no way to avoid that.
We chatted about this with Yuan and Ke a long time ago and chose to make the AddEdge do this nodearg update. It is in the following code snippet from AddEdge:
```
  if (src_arg != dst_arg) {
    if (src_arg->Type() != dst_arg->Type()) {
      // The output type of source node arg does not match the input type of destination node arg.
      ORT_THROW("Argument type mismatch when adding edge.");
    }
    *dst_arg_pointer = src_arg;
  }
```